app/controllers/coronavirus/sub_sections_controller.rb

```diff
@@ -9,11 +9,18 @@ def new
 
     def create
       @sub_section = page.sub_sections.new(sub_section_params)
-      if @sub_section.save && draft_updater.send
-        redirect_to coronavirus_page_path(page.slug), notice: "Sub-section was successfully created."
-      else
-        @sub_section.errors.add :base, draft_updater.errors.to_sentence
+
+      SubSection.transaction do
+        if @sub_section.save
+          raise ActiveRecord::Rollback unless draft_updater.send
+        end
+      end
+
+      if !@sub_section.created_at || draft_updater.errors.any?
```

Comment:
I'm a bit confused by this dual conditional here - particularly as we only test one. 

I'm assuming the two parts of it are needed because the item can be invalid or the draft update can fail, which actually makes me think we've missed something on the update method - as that will just raise an exception.

It looks like we've also now lost the ability to display model errors with `flash["alert"] = draft_updater.errors.to_sentence`.

Perhaps both the methods need a little bit of a quick pseudo code think through and to decide what to test as both seem to have hidden surprises.

Reply:
This work has made us realise that we weren't dealing with errors properly as they came from the model. We've updated to avoid the need for the conditional and to ensure that errors are caught by our model layer.

Reply:
Great stuff 👍 

It'd be good to have tests for both scenarios - since they're quite confusing and easy to overlook.

I also feel like it'd be good if the code explicitly catered for them so it was easier to follow. I wondered if something like this would work: 

```
@sub_section = page.sub_sections.new(sub_section_params)

unless @sub_section.valid?
  render :new
  return
end

SubSection.transaction do
  @sub_section.save!
  raise ActiveRecord::Rollback unless draft_updater.send
end

if draft_updater.errors
  redirect_to coronavirus_page_path(page.slug), { notice: "Sub-section was successfully created." }
else
  flash["alert"] = draft_updater.errors.to_sentence
  render :new
end
```

Reply:
Ooh this is much nicer! ❤️

Reply:
I'd not seen the `valid?` method before, this is really clean, thank you :)

Reply:
> Ooh this is much nicer! ❤️ I forget now, but I'm assuming if the call to `valid?` fails, it will display the model errors?

Oh thanks - yeah, once you've called `valid?` the model has it's `errors` property populated and will fall into this condition: https://github.com/alphagov/collections-publisher/blob/7cf7c0f4a870c3b57df531767db7e4d8d7250c6b/app/views/coronavirus/sub_sections/edit.html.erb#L25-L27

What I'm less clear on is whether setting `flash["alert"]` works as I'm not sure where that's handled (I copied it from your earlier suggestion)